src/wp-admin/includes/image.php

```diff
@@ -252,106 +254,133 @@ function wp_create_image_subsizes( $file, $attachment_id ) {
+			// scale the image and use it as the "full" size.
+			if ( $threshold && ( $image_meta['width'] > $threshold || $image_meta['height'] > $threshold ) ) {
+				$editor = wp_get_image_editor( $file, array( 'mime_type' => $output_mime_type ) );
 
-			// If there is EXIF data, rotate according to EXIF Orientation.
-			if ( ! is_wp_error( $resized ) && is_array( $exif_meta ) ) {
-				$resized = $editor->maybe_exif_rotate();
-				$rotated = $resized;
-			}
+				if ( is_wp_error( $editor ) ) {
+					// This image cannot be edited.
+					return $image_meta;
+				}
+				$editor->set_mime_type( $output_mime_type );
 
-			if ( ! is_wp_error( $resized ) ) {
-				// Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
-				// This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
-				$saved = $editor->save( $editor->generate_filename( 'scaled' ) );
+				// Resize the image.
+				$resized = $editor->resize( $threshold, $threshold );
+				$rotated = null;
 
-				if ( ! is_wp_error( $saved ) ) {
-					$image_meta = _wp_image_meta_replace_original( $saved, $file, $image_meta, $attachment_id );
+				// If there is EXIF data, rotate according to EXIF Orientation.
+				if ( ! is_wp_error( $resized ) && is_array( $exif_meta ) ) {
+					$resized = $editor->maybe_exif_rotate();
+					$rotated = $resized;
+				}
 
-					// If the image was rotated update the stored EXIF data.
-					if ( true === $rotated && ! empty( $image_meta['image_meta']['orientation'] ) ) {
-						$image_meta['image_meta']['orientation'] = 1;
+				if ( ! is_wp_error( $resized ) ) {
+					// Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
+					// This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
+					$saved = $editor->save( $editor->generate_filename( 'scaled' ), $output_mime_type );
+
+					if ( ! is_wp_error( $saved ) ) {
+						if ( 0 === $output_index ) {
+							$image_meta = _wp_image_meta_replace_original( $saved, $file, $image_meta, $attachment_id );
+						}
+						// If the image was rotated update the stored EXIF data.
+						if ( true === $rotated && ! empty( $image_meta['image_meta']['orientation'] ) ) {
+							$image_meta['image_meta']['orientation'] = 1;
+						}
+					} else {
+						// TODO: Log errors.
 					}
 				} else {
 					// TODO: Log errors.
 				}
 			} else {
-				// TODO: Log errors.
-			}
-		} elseif ( ! empty( $exif_meta['orientation'] ) && 1 !== (int) $exif_meta['orientation'] ) {
-			// Rotate the whole original image if there is EXIF data and "orientation" is not 1.
+				// Generate additional full sized images.
+				if ( $output_index > 0 ) {
+					if ( empty( $image_meta['sources'][ $output_mime_type ] ) ) {
+						$extension = wp_get_default_extension_for_mime_type( $output_mime_type );
+						$saved     = $editor->save( $editor->generate_filename( null, null, $extension ) );
```

Comment:
I found a problem with this section due to the way core works, the problem is for the full-size image which would create a different filename additional mime types besides the original mime type.

The problem basically is that additional mime types for the full-size image would contain a suffix on the file name and that would be different than the original image instead of just having a file extension different. Due the way the `generate_filename` currently works.

In order for us to avoid this, we need to update this section of code and the original function as well. By passing an empty value as suffix and making sure that the function `generate_filename` just creates a suffix only if the parameter is `null`

```suggestion
						$saved     = $editor->save( $editor->generate_filename( '', null, $extension ) );
```

Update the original function as well here:

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-image-editor.php#L429-L431

```suggestion
		if ( null === $suffix ) {
			$suffix = $this->get_suffix();
		}
```

Reply:
I noticed this as well and your suggested fix might be the best approach, I thought I had already resolved separately but will check.